miniapps/dpg/util/weakform.cpp

```diff
@@ -0,0 +1,844 @@
+         }
+      }
+      test_offs.PartialSum();
+      trial_offs.PartialSum();
+
+      G.SetSize(test_offs.Last()); G = 0.0;
+      vec.SetSize(test_offs.Last()); vec = 0.0;
+      B.SetSize(test_offs.Last(),trial_offs.Last()); B = 0.0;
+
+
+      for (int j = 0; j < test_fecols.Size(); j++)
+      {
+         int order_j = test_fecols[j]->GetOrder();
+
+         eltrans = mesh->GetElementTransformation(iel);
+         const FiniteElement & test_fe =
+            *test_fecols[j]->GetFE(eltrans->GetGeometryType(), order_j);
+
+         for (int k = 0; k < lfis[j]->Size(); k++)
+         {
+            (*lfis[j])[k]->AssembleRHSElementVect(test_fe,*eltrans,vec_e);
+            vec.AddSubVector(vec_e,test_offs[j]);
+         }
+
+         for (int i = 0; i < test_fecols.Size(); i++)
+         {
+            int order_i = test_fecols[i]->GetOrder();
+            eltrans = mesh->GetElementTransformation(iel);
+            const FiniteElement & test_fe_i =
+               *test_fecols[i]->GetFE(eltrans->GetGeometryType(), order_i);
+
+            for (int k = 0; k < test_integs(i,j)->Size(); k++)
+            {
+               if (i==j)
+               {
+                  (*test_integs(i,j))[k]->AssembleElementMatrix(test_fe,*eltrans,Ge);
+               }
+               else
+               {
+                  (*test_integs(i,j))[k]->AssembleElementMatrix2(test_fe_i,test_fe,*eltrans,
+                                                                 Ge);
+               }
+               G.AddSubMatrix(test_offs[j], test_offs[i], Ge);
+            }
+         }
+
+         for (int i = 0; i < trial_fes.Size(); i++)
+         {
+            if (IsTraceFes[i])
+            {
+               for (int k = 0; k < trial_integs(i,j)->Size(); k++)
+               {
+                  int face_dof_offs = 0;
+                  for (int ie = 0; ie < numfaces; ie++)
+                  {
+                     int iface = faces[ie];
+                     FaceElementTransformations * ftr = mesh->GetFaceElementTransformations(iface);
+                     const FiniteElement & tfe = *trial_fes[i]->GetFaceElement(iface);
+                     (*trial_integs(i,j))[k]->AssembleTraceFaceMatrix(iel,tfe,test_fe,*ftr,Be);
+                     B.AddSubMatrix(test_offs[j], trial_offs[i]+face_dof_offs, Be);
+                     face_dof_offs+=Be.Width();
+                  }
+               }
+            }
+            else
+            {
+               const FiniteElement & fe = *trial_fes[i]->GetFE(iel);
+               eltrans = mesh->GetElementTransformation(iel);
+               for (int k = 0; k < trial_integs(i,j)->Size(); k++)
+               {
+                  (*trial_integs(i,j))[k]->AssembleElementMatrix2(fe,test_fe,*eltrans,Be);
+                  B.AddSubMatrix(test_offs[j], trial_offs[i], Be);
+               }
+            }
+         }
+      }
+
+      // Form Normal Equations B^T G^-1 B = B^T G^-1 l
```

Comment:
Some high-level documentation about the equations would be more helpful in the documentation before `class DPGWeakForm` in weakform.hpp, with some description of how `G` is defined by test integrators, `B` by trial integrators, etc.

Reply:
references are given in the examples